Introduce index_type and standardize size_type across Kokkos backends

janciesko opened the pull request:
This PR introduces a signed `index_type` for each memory and execution space and standardizes `size_type` as unsigned across all Kokkos backends. This aligns Kokkos with the mdspan model where `index_type` is signed and `size_type` is unsigned.

### Motivation
- **Signed indexing**: Signed indices are safer for arithmetic (e.g., negative ranges) and avoid undefined behavior in overflow cases.
- **Consistency**: Different backends previously used different index/size semantics; this makes them uniform.
- **mdspan compatibility**: `std::mdspan` uses `index_type` (signed) and `size_type` (unsigned); Kokkos Views should mirror this.

### Changes
Memory spaces
- Add `index_type = std::make_signed_t<size_type>` to all memory spaces:
  - Cuda: `CudaSpace`, `CudaUVMSpace`, `CudaHostPinnedSpace`
  - HIP: `HIPSpace`, `HIPHostPinnedSpace`, `HIPManagedSpace`
  - Host: `HostSpace`, `AnonymousSpace`
  - OpenACC: `OpenACCSpace`
  - SYCL: `SYCLDeviceUSMSpace`, `SYCLSharedUSMSpace`, `SYCLHostUSMSpace`
Execution spaces
- Propagate `index_type` from memory space to execution spaces (Cuda, HIP, HPX, OpenACC, OpenMP, SYCL, Serial, Threads).
Policies
- Default policy `index_type` is now `execution_space::index_type` instead of `execution_space::size_type`.
- Improved overflow handling in `RangePolicy` for signed index types.
Views and CRS
- `BasicView`: Use `mdspan_type::index_type` for `index_type`.
- `ViewTraits`: Add `index_type`.
- `Crs`: Use `index_type` in transpose and count/fill utilities.

janciesko commented:
Retest this please.

dalg24 commented on core/src/HIP/Kokkos_HIP_Space.hpp:
```diff
@@ -211,7 +212,8 @@ class HIPManagedSpace {
   using device_type     = Kokkos::Device<execution_space, memory_space>;
-  using size_type       = unsigned int;
+  using index_type      = unsigned int;
+  using size_type       = index_type;
```
This one is wrong
```suggestion
  using size_type       = unsigned int;
  using index_type      = std::make_signed_t<size_type>;
```

dalg24 commented on core/src/Kokkos_AnonymousSpace.hpp:
```diff
@@ -23,6 +23,7 @@ class AnonymousSpace {
   using memory_space    = AnonymousSpace;
   using execution_space = Kokkos::DefaultExecutionSpace;
   using size_type       = size_t;
+  using index_type      = size_type;
```
```suggestion
  using index_type      = std::make_signed_t<size_type>;
```

dalg24 reviewed (dismissed):
We agreed `index_type` was going to be signed, defined as `std::make_signed_t<size_type>;`

dalg24 commented on core/src/View/Kokkos_ViewTraits.hpp:
```diff
@@ -536,7 +536,8 @@ struct ViewTraits {
 
-  using size_type = typename MemorySpace::size_type;
+  using size_type  = typename MemorySpace::size_type;
+  using index_type = typename MemorySpace::index_type;
```
Why do we want it in the View traits?
Why is it part of this PR?

cwpearson replied:
it might be surprising for `size_type` to exist and `index_type` to not exist if they're side-by-side everywhere else.

janciesko replied:
We don't have a use for it, @dalg24 is right, rolled back change.

dalg24 replied:
I am confused.  What was rolled back?  Why is this change still part of this PR?

janciesko replied:
~~Forgot about this and put it back. Reverted.~~

`[‎core/src/Kokkos_Crs.hpp]` seems to have a dependency on `traits::index_type` now.

dalg24 replied:
Why are the Crs changes part of this PR in the first place?

dalg24 replied:
Ping

janciesko replied:
Hey Damien, CRS uses ViewTraits and RangePolicy with an explicit index type; when the default policy index becomes execution_space::index_type, the CRS helpers should probably use that too.  It is for consistency with the index-type rollout.

dalg24 replied:
I do not think that we should handle Crs as part of this PR.
Please withdraw these changes.

dalg24 commented on core/src/OpenACC/Kokkos_OpenACCSpace.hpp:
```diff
@@ -26,7 +26,8 @@ class OpenACCSpace {
 
-  using size_type = size_t;
+  using size_type  = size_t;
+  using index_type = size_type;
```
```suggestion
  using index_type = std::make_signed_t<size_type>;
```

dalg24 commented on core/src/Kokkos_HostSpace.hpp:
```diff
@@ -38,6 +38,7 @@ class HostSpace {
   //! Tag this class as a kokkos memory space
   using memory_space = HostSpace;
   using size_type    = size_t;
+  using index_type   = size_type;
```
```suggestion
  using index_type   = std::make_signed_t<size_type>;
```

cwpearson reviewed (commented):
Looks like we refer to `size_type` in the docs as what we now mean by `index_type`?

https://github.com/kokkos/kokkos-core-wiki/blob/08d3bfd041455efa4154290a578a707f13e15ae9/docs/source/API/core/view/view.rst?plain=1#L191-L193

(that's not the only place, e.g. DynRankView)

masterleinad commented:
> Looks like we refer to `size_type` in the docs as what we now mean by `index_type`?

Yes, the idea is to replace `size_type` with `index_type` and possibly deprecate `size_type`.

janciesko commented:
@masterleinad, I've added two separate tests but simplified the `check_conversion_safety`.

masterleinad commented:
> @masterleinad, I've added two separate tests but simplified the `check_conversion_safety`.

This is failing on pretty much every single CI build. Does the tests pass for you locally?

janciesko commented:
Fixing

janciesko committed:
Add index_type and standardize size_type across backends

Add signed index_type to all memory/execution spaces, use it as the default
for policies, and update Views/CRS. Aligns with mdspan index/size type model.

Signed-off-by: Jan Ciesko <jan.ciesko@gmail.com>

dalg24 commented on core/src/impl/Kokkos_HostSpace_deepcopy.cpp:
```diff
@@ -128,8 +128,7 @@ template <typename ExecutionSpace>
   constexpr uint8_t z_u8 = 0x00;
-  using policy_t =
-      Kokkos::RangePolicy<ExecutionSpace, Kokkos::IndexType<size_t>>;
+  using policy_t         = Kokkos::RangePolicy<ExecutionSpace>;
```
This would break if we changed the definition of `HostSpace::index_type`.
Any particular reason to touch this code?

janciesko replied:
It would, hence the index type is part of the exec space. We cannot arbitrarily change that.

dalg24 replied:
I don't understand the answer.  I am saying was there a particular harm in being explicit?  Is it better to have that code work implicitly because the default index type is (hopefully) going to be large enough.

janciesko replied:
"Being explicit" and "index type is (hopefully) going to be large enough" are good considerations but I think they could be addressed when defining the default `index_type` (which could be size_t) for that exec space.

masterleinad replied:
No, it's not. We explicitly want the index type here tp be `size_t` since that's the maximum range of the allocation. `IndexType` neesd to be kernel-specific (with a default as `ExecutionSpace::index_type`).

janciesko replied:
aah, I mixed up the words. Yes, `size_type` here which is potentially larger than index_type. Good point, thanks Daniel.

janciesko replied:
Probably not even `size_type` but `size_t`.

dalg24 commented on core/src/View/Kokkos_BasicView.hpp:
```diff
@@ -136,15 +136,13 @@ class BasicView {
+  using mapping_type     = typename mdspan_type::mapping_type;
+  using element_type     = typename mdspan_type::element_type;
+  using value_type       = typename mdspan_type::value_type;
+  using index_type       = typename mdspan_type::index_type;
```
Why does that change belong to this PR?

janciesko replied:
It seems reasonable to 
`using index_type = typename mdspan_type::size_type;` -> 
`using index_type = typename mdspan_type::index_type;`

crtrott replied:
I think this is fine, right now it's anyway hardcoded to size_t and it falls into the cleanup of index_type being consistently used.

dalg24 replied:
Again I do not understand why this change belong here.  Please revert it and propose it as a follow up PR

masterleinad commented on core/src/impl/Kokkos_HostSpace_deepcopy.cpp:
```diff
@@ -128,8 +128,7 @@ template <typename ExecutionSpace>
 void hostspace_parallel_zeromemset(const ExecutionSpace& exec, void* dst,
                                    size_t n) {
   constexpr uint8_t z_u8 = 0x00;
-  using policy_t =
-      Kokkos::RangePolicy<ExecutionSpace, Kokkos::IndexType<size_t>>;
```
Why did you change?

masterleinad commented on core/src/traits/Kokkos_IndexTypeTrait.hpp:
```diff
@@ -47,7 +47,7 @@ struct IndexTypePolicyMixin<Kokkos::IndexType<IntegralIndexType>,
   static constexpr auto show_index_type_error_in_compilation_message =
       show_extra_index_type_erroneously_given_to_execution_policy<
           std::conditional_t<base_t::index_type_is_defaulted, void,
-                             typename base_t::schedule_type>>{};
```
That was just wrong it seems.

janciesko replied:
I think so

dalg24 commented on core/src/View/Kokkos_ViewAlloc.hpp:
```diff
@@ -84,9 +87,9 @@ struct ViewValueFunctor {
-    PolicyType policy(space, 0, n);
+    using index_type = typename ExecSpace::index_type;
+    using PolicyType = Kokkos::RangePolicy<ExecSpace, index_type, Tag>;
+    PolicyType policy(space, index_type(0), index_type(n));
```
The original code was explicitly requesting 64-bit integers.
Why are we changing it to something that is backend-specific?
Isn't going to fall over say with a 9GB allocation of some 4-byte non-trivially constructible user-defined type (e.g. `n > INT_MAX` on CUDA or HIP).

janciesko replied:
This is an interesting case. Looking for more input if we should hard-code such loops to use 64 bits indices or infer from index_type.

janciesko replied:
Same rational applies here I think, thanks for pointing it out Damien. Still we likely want to use `size_t` as the index type.

masterleinad replied:
Generally, you should probaby leave `size_t` alone (if it's not the alias declaration `using size_type = size_t`).

crtrott replied:
Same needs to be reverted since we may have allocations of more than 2B elements.

janciesko replied:
Agreed. Reverted in: https://github.com/kokkos/kokkos/pull/7441/changes/53b132f5df4492a1456115bc5ea74121c1734ccb

dalg24 commented on core/src/View/Kokkos_ViewAlloc.hpp:
```diff
@@ -40,11 +40,12 @@ struct ViewValueFunctor {
   KOKKOS_FUNCTION
       std::enable_if_t<std::is_default_constructible_v<SameValueType>>
-      operator()(ConstructTag, const size_t i) const {
+      operator()(ConstructTag, const ExecSpace::index_type i) const {
```
It looks like we were explicitly requesting 64-bit integers to index into that array we are allocating.

janciesko replied:
For consistency we should probably use index type here.

masterleinad replied:
You want to use the `index_type` of the policy here but since that isn't available we need tp be explicit and use `size_t`.

janciesko replied:
Agreed, reverted, thanks for the rational.

crtrott replied:
It's not yet reverted is it? And the rational for using 64bit indexing in all these guys is that we are instantiating these kernels to loop over the entries of a View during initialization and destruction. This thus needs to cover the range of the index_type we are using in the View, which currently is actually 64bit. And it needs to be to cover more than 2B elements. Since we are not passing the View in here lets leave these alone for now, and rework this when we actually can change the View index_type, thus we could know here at compile time that we wouldn't have more than 2B iterations.

dalg24 commented on core/src/Kokkos_ExecPolicy.hpp:
```diff
@@ -280,9 +280,17 @@ class RangePolicy : public Impl::PolicyTraits<Properties...> {
-                               std::numeric_limits<member_type>::max()));
+        if constexpr (std::is_signed_v<member_type>) {
+          // avoid overflow warnings by checking the size of the types e.g.
+          // conversion ‘long int’ to ‘int’ changes value
+          // from ‘9223372036854775807’ to ‘-1’
+          if constexpr (sizeof(member_type) <= sizeof(IndexType))
+            warn |= (bound > static_cast<IndexType>(
+                                 std::numeric_limits<member_type>::max()));
+          else {
+            warn |= (bound > std::numeric_limits<IndexType>::max());
+          }
+        }
```
Was there a gap in our RangePolicy constructor testing?

janciesko replied:
See comment in-line

dalg24 replied:
I can see the comment, I am asking why we did not realize this before and whever we need to update `core/unit_test/TestRangePolicyConstructors.hpp`

janciesko replied:
We did not realize this before as our tests tested for positive (or zero) indices only. Testing with negative indices produces those warnings now. The new tests are in `TestRangePolicyConstructors.hpp`.

ldh4 replied:
I may be little confused on the provided comments,
but because of this condition in the check:
```
        if constexpr (std::is_signed_v<IndexType> !=
                      std::is_signed_v<member_type>) {
```
this portion of check is for when signedness of `IndexType` and `member_type` don't match (except when IndexType is a user-defined type)
The unit tests you added in `TestRangePolicyConstructors.hpp` seems to be only signed (`long long`) to signed (`int`) conversion.

janciesko replied:
Yes, the `long long` + `RangePolicy<int>` death tests look like “signed to signed”; note that we're testing the round-trip / narrowing path there. We test for the signedness mismatch between the bound type and the policy index type in the `range_policy_implicitly_converted_bounds` tests.

dalg24 commented on core/src/traits/Kokkos_IndexTypeTrait.hpp:
```diff
@@ -47,7 +47,7 @@ struct IndexTypePolicyMixin<Kokkos::IndexType<IntegralIndexType>,
       show_extra_index_type_erroneously_given_to_execution_policy<
           std::conditional_t<base_t::index_type_is_defaulted, void,
-                             typename base_t::schedule_type>>{};
+                             typename base_t::index_type>>{};
```
So if I understand correctly, what was wrong was the error message we would produce, if specifying twice the integer type to use with the policy (via template arguments)
https://godbolt.org/z/raGMTsxrd

masterleinad replied:
Could also be in a separate pull request.

janciesko committed:
More updates
* Add index_type to traits for its use in Kokkos_Crs
* Improve comment in check_conversion_safety
* Add the threadsafe flag

Signed-off-by: Jan Ciesko <jan.ciesko@gmail.com>

crtrott commented on core/src/Cuda/Kokkos_CudaSpace.hpp:
```diff
@@ -49,8 +49,8 @@ class CudaSpace {
   using device_type     = Kokkos::Device<execution_space, memory_space>;
-
-  using size_type = unsigned int;
+  using size_type       = unsigned int;
+  using index_type      = std::make_signed_t<size_type>;
```
Should we turn the logic around? mdspan defines index_type and then makes size_type the unsigned version. I do agree that we should have a default index type for the execution spaces which is signed, in order to support negative indices.

crtrott replied:
Note this is a question, not a mandate to change. Ending up with `unsigned int` / `int` respectively for our GPU spaces and the 64bit variants for host is fine with me.

dalg24 replied:
You mean this
```suggestion
  using index_type      = int;
  using size_type       = std::make_unsigned_t<index_type>;
```

Fine by me.

crtrott commented on core/src/View/Kokkos_ViewLegacy.hpp:
```diff
@@ -323,7 +323,7 @@ class View : public ViewTraits<DataType, Properties...> {
   using element_type = typename traits::value_type;
   // using value_type -> conflicts with traits::value_type
-  using index_type = typename traits::memory_space::size_type;
+  using index_type = typename traits::index_type;
```
Let's leave this guy alone.

crtrott commented on core/src/Kokkos_ExecPolicy.hpp:
```diff
@@ -248,20 +248,25 @@ class RangePolicy : public Impl::PolicyTraits<Properties...> {
+          // check unsigned to signed
+          if constexpr (std::is_signed_v<member_type>) {
+            // avoid overflow warnings by checking the size of the types e.g.
+            // conversion ‘long int’ to ‘int’ changes value
```
```suggestion
            // conversion ‘unsigned int’ to ‘int’ changes value
```

crtrott replied:
We are in the IndexType is unsigned case right now.

crtrott reviewed (changes requested):
Couple comments to be addressed. But there is also one bigger one: we are changing the sign of RangePolicy begin/end. This might be a bit worrisome, and will pop up warnings in downstream code everywhere. On the other hand I am also feeling really uncomfortable with negative begin values not working by default. Thoughts?

janciesko committed:
Fix RangePolicy conversion check and restore View index_type

Signed-off-by: Jan Ciesko <jan.ciesko@gmail.com>
Co-authored-by: Christian Trott <crtrott@sandia.gov>

janciesko committed:
Use size_t RangePolicy for ViewValueFunctor, fixes

Signed-off-by: Jan Ciesko <jan.ciesko@gmail.com>

masterleinad commented on core/unit_test/TestRangePolicyConstructors.hpp:
```diff
@@ -169,12 +195,16 @@ TEST(TEST_CATEGORY_DEATH, range_policy_implicitly_converted_bounds) {
+    ASSERT_DEATH({ (void)IntPolicy(0, test_val); },
+                 get_error_msg(expected, test_val));
+
+    ASSERT_DEATH({ (void)DefaultPolicy(0, test_val); },
```
`DefaultPolicy` isn't defined.

masterleinad commented:
> But there is also one bigger one: we are changing the sign of RangePolicy begin/end. This might be a bit worrisome, and will pop up warnings in downstream code everywhere. On the other hand I am also feeling really uncomfortable with negative begin values not working by default. Thoughts?

Since `index_type` didn't exist before, I don't think (`-Wsign-compare`-type) warnings are a concern downstream generally. The change that appears in user programs is assgnment between values of different signedness but I'm not aware that the major compilers generally warn about that. Compiling with `-Wsign-conversion` in `Kokkos` would already give a lot of extra warnings.

For  the `SYCL` backend, we had used `size_type=int` and that triggered warnings in user codes whenever `ExecutionSpace::size_type` was explicitly used for a loop variable (or similar). IIn https://github.com/kokkos/kokkos/pull/8542, we changed that type to be unsigned for consistency with other backends but I haven't heard anyone complaining about more warnings.

masterleinad commented on core/src/impl/Kokkos_HostSpace_deepcopy.cpp:
```diff
@@ -147,9 +147,10 @@ void hostspace_parallel_zeromemset(const ExecutionSpace& exec, void* dst,
   Kokkos::parallel_for("Kokkos::Impl::hostspace_parallel_zeromemset",
                        policy_t(exec, 0, cnt),
-                       [=](const size_t i) { dst_p[i] = z_u64; });
+                       [=](const index_t i) { dst_p[i] = z_u64; });
```
This still wrong. There are no changes necessary in this file.

janciesko replied:
Oversight probably, thanks! Rolled back.

masterleinad commented on core/src/View/Kokkos_ViewAlloc.hpp:
```diff
@@ -84,9 +84,11 @@ struct ViewValueFunctor {
     using PolicyType =
-        Kokkos::RangePolicy<ExecSpace, Kokkos::IndexType<int64_t>, Tag>;
-    PolicyType policy(space, 0, n);
+        Kokkos::RangePolicy<ExecSpace, Kokkos::IndexType<size_t>, Tag>;
+    PolicyType policy(space, size_t(0), n);
```
Why did you change?

janciesko replied:
This seems like an oversight in the original code.

masterleinad replied:
I'd be suprised if you can create an allocation larger than `std::numeric_limits<std::int64_t>::max()`. 🙂 

Note that `size_t` isn't necessarily a 64-bit type so the new comment seems to apply better to the old code. If you keep the change here, I would remove the comment.

dalg24 replied:
I agree with Daniel, I don't see any good reason for this change.
On second thought I don't really care to revert.
I don't find the comment particularly useful.

masterleinad commented on core/unit_test/view/TestBasicView.hpp:
```diff
@@ -18,7 +18,9 @@ namespace {
-  return Kokkos::RangePolicy<ExecutionSpace>{0, extents.extent(0)};
+  using index_type = typename Kokkos::RangePolicy<ExecutionSpace>::index_type;
+  return Kokkos::RangePolicy<ExecutionSpace>{
+      0, static_cast<index_type>(extents.extent(0))};
```
Why did you change?

janciesko replied:
This is unnecessary. Rolled back.

dalg24 commented on core/unit_test/TestTeamPolicyCTAD.cpp:
```diff
@@ -19,6 +19,7 @@ struct TestTeamPolicyCTAD {
   struct SomeExecutionSpace {
     using execution_space = SomeExecutionSpace;
     using size_type       = size_t;
+    using index_type      = std::make_signed_t<size_type>;
```
```suggestion
    using index_type      = int;
```

dalg24 commented on core/unit_test/TestRangePolicyCTAD.cpp:
```diff
@@ -15,6 +15,7 @@ struct TestRangePolicyCTAD {
   struct SomeExecutionSpace {
     using execution_space = SomeExecutionSpace;
     using size_type       = size_t;
+    using index_type      = std::make_signed_t<size_type>;
```
```suggestion
    using index_type      = int;
```

dalg24 commented on core/unit_test/TestMDRangePolicyCTAD.cpp:
```diff
@@ -21,6 +21,7 @@ struct TestMDRangePolicyCTAD {
   struct SomeExecutionSpace {
     using execution_space = SomeExecutionSpace;
     using size_type       = size_t;
+    using index_type      = std::make_signed_t<size_type>;
```
```suggestion
    using index_type      = int;
```

dalg24 commented on core/unit_test/TestRange.hpp:
```diff
@@ -145,8 +180,15 @@ struct TestRange {
+
+    Kokkos::parallel_reduce(
+        Kokkos::RangePolicy<ExecSpace, ScheduleType, NegOffsetTag>(
+            (-1) * offset, N - offset),
```
```suggestion
            -offset, N - offset),
```

dalg24 commented on core/unit_test/TestRangePolicyConstructors.hpp:
```diff
@@ -18,9 +18,10 @@ import kokkos.core_impl;
+  using Policy = Kokkos::RangePolicy<>;
+  using Index  = Policy::index_type;
+
+  Index work_begin = -5;
```
Why change this unit test?

janciesko replied:
I think we can reuse this test, just extending the ranges.

dalg24 reviewed (commented):
Don't forget to also update the NextSilicon backend

janciesko committed:
Add index_type to NextSilicon backend, fixes

Signed-off-by: Jan Ciesko <jan.ciesko@gmail.com>
Co-authored-by: Damien Lebrun-Grandie <dalg24@gmail.com>

dalg24 commented on core/unit_test/TestTeamPolicyCTAD.cpp:
```diff
@@ -19,6 +19,7 @@ struct TestTeamPolicyCTAD {
   struct SomeExecutionSpace {
     using execution_space = SomeExecutionSpace;
     using size_type       = size_t;
```
```suggestion
```

dalg24 commented on core/unit_test/TestRangePolicyCTAD.cpp:
```diff
@@ -15,6 +15,7 @@ struct TestRangePolicyCTAD {
   struct SomeExecutionSpace {
     using execution_space = SomeExecutionSpace;
     using size_type       = size_t;
```
```suggestion
```

dalg24 commented on core/unit_test/TestMDRangePolicyCTAD.cpp:
```diff
@@ -21,6 +21,7 @@ struct TestMDRangePolicyCTAD {
   struct SomeExecutionSpace {
     using execution_space = SomeExecutionSpace;
     using size_type       = size_t;
```
```suggestion
```

ldh4 commented on core/src/Kokkos_ExecPolicy.hpp:
```diff
@@ -247,20 +247,25 @@ class ImplRangePolicy<ExecSpace, Properties...>
+            // avoid overflow warnings by checking the size of the types e.g.
+            // conversion ‘unsigned int’ to ‘int’ changes value
+            // from ‘4294967295’ to ‘-1’ (when IndexType is narrower)
+            if constexpr (sizeof(member_type) <= sizeof(IndexType))
```
If this produced an overflow warning when `IndexType` is narrower than `member_type`, wouldn't there be a same warning produced from L270, which does the same casting check?

Also, I'm not sure if relying on sizeof is safe, since IndexType could be any user defined type (could be implicitly convertible to and from `member_type` but may not necessarily an arithmetic type)

janciesko replied:
The `sizeof` split is only trying to avoid the specific diagnostic error you get from  
`bound > static_cast<IndexType>(std::numeric_limits<member_type>::max())` when `IndexType` is narrower than `member_type` (e.g. `member_type = unsigned long long`, `IndexType = int`): the cast of `max()` to a narrower signed type is what triggers the “value changes on conversion” style warning. 

The conversion is safe given we're in the `if constexpr (std::is_arithmetic_v<member_type>)`  branch.

ldh4 replied:
I get that this is to avoid the overflow warning and that adding this additional constraint doesn't alter the upper bound check.
However, I have a tangential concern on the usage of `sizeof`. This condition is fine when `IndexType` is an arithmetic type but not when `IndexType` is a UDT padded with extra variables.

For instance, in the test that was disabled in this PR `range_policy_round_trip_conversion_fires`, padding the struct `W` with an extra `long` will satisfy the condition `(sizeof(member_type) <= sizeof(IndexType))` even though the actual used index type is `int`. So `sizeof` seems unreliable here when `IndexType` isn't arithmetic.

janciesko replied:
Aah, yes, that's right, I have not thought that through. The entire block should be replaceable with `std::in_range`.

janciesko committed:
Revert Kokkos_Crs and ViewTraits.hpp index_type

Signed-off-by: Jan Ciesko <jan.ciesko@gmail.com>

ldh4 commented on core/unit_test/TestRangePolicyConstructors.hpp:
```diff
@@ -100,11 +124,13 @@ TEST(TEST_CATEGORY_DEATH, range_policy_round_trip_conversion_fires) {
   static_assert(std::is_convertible_v<W, Policy::index_type>);
   static_assert(std::is_convertible_v<Policy::index_type, W>);
 
-  int const n = 1;
-  [[maybe_unused]] std::string msg =
-      "Kokkos::RangePolicy bound type error: an unsafe implicit conversion is "
-      "performed";
-  ASSERT_DEATH((void)Policy(0, W(&n)), msg);
+  // FIXME: The following ASSERT_DEATH was disabled because the round-trip
+  // conversion check does not trigger with the current check_conversion_safety
+  // logic. Re-enable once the expected failure is understood and implemented.
+  // [[maybe_unused]] std::string msg =
+  //     "Kokkos::RangePolicy bound type error: an unsafe implicit conversion is
+  //     " "performed";
+  // ASSERT_DEATH((void)Policy(0, W(&n)), msg);
```
I would uncomment this and disable the test by prepending the test name with `DISABLED` (`DISABLED_range_policy_round_trip_conversion_fires`).
Or, place an extra `long` var in the struct `W`, and it will trigger the expected abort.

janciesko committed:
Roll back use of in_range / keep previous test logic. Fix some warnings.

Signed-off-by: Jan Ciesko <jan.ciesko@gmail.com>

janciesko commented on core/src/Kokkos_ExecPolicy.hpp:
```diff
@@ -240,37 +240,35 @@ class ImplRangePolicy<ExecSpace, Properties...>
 
+      // For arithmetic member_type and IndexType, use std::in_range,
+      // use the round-trip check below for UDT
       if constexpr (std::is_arithmetic_v<member_type> &&
```
@ldh4 How's this?

ldh4 replied:
Yeah maybe this is how it'll have to be, bypassing this check if `IndexType` is a UDT. I wasn't sure if there's a way to keep the bounds check correct while considering all cases of UDTs.

As for `std::in_range`, I believe it only works for integral types. We could definitely check with `std::in_range` if both `member_type` and `IndexType` are integrals. Otherwise, maybe fallback to the previous implementation (given that both types are still arithmetic types)?

janciesko replied:
`std::in_range` does not buy us anything. Reverted.

nmm0 commented on core/src/OpenACC/Kokkos_OpenACCSpace.hpp:
```diff
@@ -26,7 +26,8 @@ class OpenACCSpace {
 
-  using size_type = size_t;
+  using size_type  = size_t;
+  using index_type = std::make_signed_t<size_type>;
```
I think this is okay, but worth mentioning that this will likely be a signed 64 bit int instead of the other device execution spaces which are signed 32 bit ints

janciesko replied:
The reduction seems to be broken

janciesko replied:
@nmm0 Should be fixed.

nmm0 commented on core/src/View/Kokkos_ViewAlloc.hpp:
```diff
@@ -84,9 +84,11 @@ struct ViewValueFunctor {
     using PolicyType =
-        Kokkos::RangePolicy<ExecSpace, Kokkos::IndexType<int64_t>, Tag>;
-    PolicyType policy(space, 0, n);
+        Kokkos::RangePolicy<ExecSpace, Kokkos::IndexType<size_t>, Tag>;
```
This may or may not be 64 bit, I'd rather not driveby change this

crtrott replied:
there is a conversion thing otherwise going on, because the functor called already was using size_t.

nmm0 commented on core/src/Kokkos_ExecPolicy.hpp:
```diff
@@ -240,37 +240,44 @@ class ImplRangePolicy<ExecSpace, Properties...>
-        if constexpr (std::is_signed_v<member_type>)
-          error |= (bound > static_cast<IndexType>(
-                                std::numeric_limits<member_type>::max()));
+                    std::is_arithmetic_v<IndexType>) {
```
If I understand this correctly, and this doesn't have to work on device we may be able to simplify the representability part with https://en.cppreference.com/cpp/utility/in_range

nmm0 reviewed (commented):
First pass, I will take another look tomorrow

crtrott commented on core/src/View/Kokkos_ViewAlloc.hpp:
```diff
@@ -84,9 +84,11 @@ struct ViewValueFunctor {
   void parallel_for_implementation() {
+    // Use a 64-bit index range: allocations can exceed what the execution
+    // space default index_type (e.g. 32-bit on CUDA/HIP) can represent.
     using PolicyType =
-        Kokkos::RangePolicy<ExecSpace, Kokkos::IndexType<int64_t>, Tag>;
-    PolicyType policy(space, 0, n);
+        Kokkos::RangePolicy<ExecSpace, Kokkos::IndexType<size_t>, Tag>;
+    PolicyType policy(space, size_t(0), n);
```
its fine, only on 32bit machines would it not be 64bit but then we don't need it either.

crtrott committed:
Address two smaller review finding

Signed-off-by: Christian Trott <crtrott@sandia.gov>

crtrott commented:
Retest this please.

janciesko committed:
Use std::size_t for custom range-reduce counters

Signed-off-by: Jan Ciesko <jan.ciesko@gmail.com>

masterleinad commented on core/src/OpenACC/Kokkos_OpenACC_ParallelReduce_Range.hpp:
```diff
@@ -171,8 +171,8 @@ void OpenACCParallelReduceCustom(Schedule<Static>, Policy const& apolicy,
-    IndexType tSize = chunk_size;
-    IndexType tStep;
+    std::size_t tSize = static_cast<std::size_t>(chunk_size);
+    std::size_t tStep;
```
What's going on here?

janciesko replied:
For OpenACC, the reduction was producing wrong results in our unit tests with the defaulted `index_type` which now is a `signed` type. 

It seems that what the compiler produces for those binary operations in that reduction is wrong. Given that `chunk_size` is always positive, I am not sure which particular binary op causes this.

Casting to an unsigned type resolved this issue. This applies only to the OpenACC backend.

masterleinad replied:
@seyonglee Can you provide more insight into this?

seyonglee replied:
I don't know why but sometimes explicit `acc loop seq` directive causes incorrect code generation; deleting them will fix the error.

janciesko replied:
This seems to have fixed it. Great to see this collaboration. Thanks again @seyonglee @masterleinad

seyonglee committed:
Fix a bug in parallel_reduce with range policy

Signed-off-by: Seyong Lee <lees2@ornl.gov>

janciesko commented:
I think this is ready to go. @masterleinad, @seyonglee could you finalize the review?
